ethereum/p2p/src/main/java/tech/pegasys/pantheon/ethereum/p2p/discovery/internal/PeerDiscoveryController.java

```diff
@@ -231,6 +282,30 @@ private boolean whitelistIfPresentIsNodePermitted(final DiscoveryPeer sender) {
         .orElse(true);
   }
 
+  private synchronized void handleBlockAddedEvent(
+      final BlockAddedEvent event, final Blockchain blockchain) {
+    final Collection<DiscoveryPeer> activePeers = peerTable.getAllPeers();
+
+    LOG.trace(
+        "Handling block added event - will check if {} active peers are still permitted (blockHash = {})",
+        activePeers.size(),
+        event.getBlock().getHash());
+
+    activePeers.stream().filter(p -> !isPeerPermitted(p)).forEach(this::dropFromPeerTable);
+  }
+
+  private boolean isPeerPermitted(final Peer peer) {
+    final EnodeURL localPeerEnodeURL = new EnodeURL(localPeer.getEnodeURI());
+    final EnodeURL remotePeerEnodeURL = new EnodeURL(peer.getEnodeURI());
+
+    return nodePermissioningController
+        .map(
+            c ->
+                c.isPermitted(localPeerEnodeURL, remotePeerEnodeURL)
+                    || c.isPermitted(remotePeerEnodeURL, localPeerEnodeURL))
```

Comment:
Do we want to say that the peer is permitted only when communication in both ways is permitted?